network.go

```diff
@@ -0,0 +1,365 @@
+
+	NetworkPairs []NetworkInterfacePair
+}
+
+// NetworkType describes the type of network specification.
+type NetworkType string
+
+const (
+	// NoopNetworkType is the No-Op network.
+	NoopNetworkType NetworkType = "noop"
+
+	// CNINetworkType is the CNI network.
+	CNINetworkType NetworkType = "cni"
+)
+
+// Set sets a network type based on the input string.
+func (networkType *NetworkType) Set(value string) error {
+	switch value {
+	case "noop":
+		*networkType = NoopNetworkType
+		return nil
+	case "cni":
+		*networkType = CNINetworkType
+		return nil
+	default:
+		return fmt.Errorf("Unknown network type %s", value)
+	}
+}
+
+// String converts a network type to a string.
+func (networkType *NetworkType) String() string {
+	switch *networkType {
+	case NoopNetworkType:
+		return string(NoopNetworkType)
+	case CNINetworkType:
+		return string(CNINetworkType)
+	default:
+		return ""
+	}
+}
+
+// newNetwork returns a network from a network type.
+func newNetwork(networkType NetworkType) network {
+	switch networkType {
+	case NoopNetworkType:
+		return &noopNetwork{}
+	case CNINetworkType:
+		return &cni{}
+	default:
+		return &noopNetwork{}
+	}
+}
+
+// genRandomInterface generates locally a unique interface name
+func genRandomInterface(prefix string) (string, error) {
+	var random *rand.Rand
+	for i := 0; i < 100; i++ {
+		iface := fmt.Sprintf("%s_%x", prefix, random.Uint32())
+
+		_, err := netlink.LinkByName(iface)
+		if err != nil {
+			return iface, nil
+		}
+	}
+
+	// The chances of the failure are remote
+	return "", fmt.Errorf("unable to create unique interface name")
+}
+
+func createTAPInterface() (*NetworkInterface, error) {
+	linkName, err := genRandomInterface("tap")
+	if err != nil {
+		return nil, err
+	}
+
+	tap := &netlink.Tuntap{
+		LinkAttrs: netlink.LinkAttrs{Name: linkName},
+		Mode:      netlink.TUNTAP_MODE_TAP,
+	}
+
+	err = netlink.LinkAdd(tap)
+	if err != nil {
+		return nil, fmt.Errorf("link add %s %v", linkName, err)
+	}
+
+	link, err := netlink.LinkByName(linkName)
+	if err != nil {
+		return nil, fmt.Errorf("LinkByName %s %v", linkName, err)
+	}
+
+	_, ok := link.(*netlink.Tuntap)
+	if !ok {
+		return nil, fmt.Errorf("incorrect interface type %s %v", linkName, link)
+	}
+
+	tapInterface := &NetworkInterface{
+		Name: linkName,
+	}
+
+	return tapInterface, nil
+}
+
+func deleteTAPInterface(tapInterface *NetworkInterface) error {
+	link, err := netlink.LinkByName(tapInterface.Name)
+	if err != nil {
+		return fmt.Errorf("LinkByName %s %v", tapInterface.Name, err)
+	}
+
+	tapl, ok := link.(*netlink.Tuntap)
+	if !ok {
+		return fmt.Errorf("incorrect interface type %s %v", tapInterface.Name, link)
+	}
+
+	if tapl == nil || tapl.Index == 0 {
+		return fmt.Errorf("delete tap interface not initialized")
+	}
+
+	err = netlink.LinkDel(tapl)
+	if err != nil {
+		return fmt.Errorf("destroy bridge %v", err)
+	}
+
+	*tapInterface = NetworkInterface{}
+
+	return nil
+}
+
+func createBridge() (*netlink.Bridge, string, error) {
```

Comment:
Could we use [libsnnet](https://github.com/01org/ciao/tree/master/networking/libsnnet) here ?